ansible_collections/f5networks/f5_modules/tests/unit/modules/network/f5/test_bigip_device_license.py

```diff
@@ -102,3 +105,75 @@ def test_create(self, *args):
 
         results = mm.exec_module()
         assert results['changed'] is True
+
+    def test_renewal(self, *args):
+        set_module_args(
+            dict(
+                license_key='xxxx-yyyy-zzzz',
+                license_server='foo-license.f5.com',
+                accept_eula=True,
+                state='latest',
+                provider=dict(
+                    server='localhost',
+                    password='password',
+                    user='admin'
+                )
+            )
+        )
+
+        module = AnsibleModule(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            required_if=self.spec.required_if
+        )
+        mm = ModuleManager(module=module)
+
+        # Override methods to force specific logic in the module to happen
+        mm.exists = Mock(side_effect=[True, True])
+        mm.is_revoked = Mock(return_value=False)
+        mm.license_valid = Mock(return_value=False)
+        mm.read_dossier_from_device = Mock(return_value=True)
+        mm.generate_license_from_remote = Mock(return_value=True)
+        mm.upload_license_to_device = Mock(return_value=True)
+        mm.upload_eula_to_device = Mock(return_value=True)
+        mm.reload_license = Mock(return_value=True)
+        mm._is_mcpd_ready_on_device = Mock(return_value=True)
+
+        results = mm.exec_module()
+        assert results['changed'] is True
+
+    def test_no_renewal(self, *args):
+        set_module_args(
+            dict(
+                license_key='xxxx-yyyy-zzzz',
+                license_server='foo-license.f5.com',
+                accept_eula=True,
+                state='latest',
+                provider=dict(
+                    server='localhost',
+                    password='password',
+                    user='admin'
+                )
+            )
+        )
+
+        module = AnsibleModule(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            required_if=self.spec.required_if
+        )
+        mm = ModuleManager(module=module)
+
+        # Override methods to force specific logic in the module to happen
+        mm.exists = Mock(side_effect=[True, True])
+        mm.is_revoked = Mock(return_value=False)
+        mm.license_valid = Mock(return_value=True)
+        mm.read_dossier_from_device = Mock(return_value=True)
+        mm.generate_license_from_remote = Mock(return_value=True)
+        mm.upload_license_to_device = Mock(return_value=True)
+        mm.upload_eula_to_device = Mock(return_value=True)
+        mm.reload_license = Mock(return_value=True)
+        mm._is_mcpd_ready_on_device = Mock(return_value=True)
+
+        results = mm.exec_module()
+        assert results['changed'] is False
```

Comment:
I was under the impression you would use the jsons to simulate responses from device and test that way but ok

Reply:
Yes - I also thought to use JSON payloads for mocking HTTP response from BIGIP REST API  (i.e. in JS, we use nock module for this: https://github.com/nock/nock)
however it appears to be impossible to do such testing with the current setup.

Reply:
it is not impossible look at other unit tests you will see how it is done

Reply:
you can patch F5RestClient, if you are not very intimate with Mock it could be tricky as you need to mock the client instance and then its 'api' property mock to return another mock (the session_mock) and then within the session_mock you return value of .get, but ok lets leave it for now.